### PR TITLE
Add sequence_loss and sequence_loss_by_example working with dynamic_rnn without having to change outputs shape

### PR DESCRIPTION
**Pull request by chenghuige**

### CODE REVIEW

**Comment by chenghuige** [comment]
I signed it!

**Comment by drpngx** [comment]
@lukaszkaiser LMK if you have cycles to handle this.

**Comment by lukaszkaiser** [comment]
Sorry, I do not. I think Thang Living is already working on it. I'll
reassign when I get back to a laptop, or you do it. Thanks!

Lukasz

**Comment by drpngx** [comment]
No worries, thanks!

**Comment by lmthang** [comment]
Hi @drpngx,

Eugene @ebrevdo is revising the dynamic_rnn_decoder and I believe he wants to hold off all developments involving dynamic_rnn_decoder for the moment.

PS: while I'll try my best to provide as much inputs as I can, it's probably better to assign these requests to a software engineer for timely responses.

-Thang

**Comment by drpngx** [comment]
OK, let's wait to hear back from @ebrevdo .

**Comment by drpngx** [comment]
@lmthang please pull rebase & push again.

**Comment by drpngx** [comment]
Sorry @chenghuige could you pull rebase and push again?

**Comment by rmlarsen** [comment]
@lmthang please review this.

**Comment by ebrevdo** [comment]
So this calculates the log perplexity?  why is it called sequence_loss and
sequence_loss_by_example?

**Comment by vrv** [comment]
It looks like there are conflicts -- @chenghuige can you address the conflicts and also @ebrevdo's most recent question?  Thanks!

**Comment by vrv** [comment]
Haven't heard back in almost a month so I'll close for now -- please send a new PR when you get around to addressing the comments, thank you!

**Comment by mhnatiuk** [comment]
Hi,
I'm trying to use tf.contrib.seq2seq.sequence_loss with dynamic_decode output, but I always get tensor incompatibility error, because, the output of dynamic_decoder is variable, and input to sequence_loss is not. Basically, example code that raises this problem is:


    `with tf.device("/cpu:0"), tf.name_scope("embeddings"):
            W_target_emb = tf.Variable(tf.random_uniform([target_vocab_size, size], -1.0, 1.0), name="W_target_emb")
            target_embedded_chars = tf.nn.embedding_lookup(W_target_emb, tf.stack(self.decoder_inputs))
        #target_embedded_chars = tf.unstack(target_embedded_chars)
        with tf.name_scope("decoder"):
            half = tf.constant(0.5)
            dec_inp = tf.cast(tf.stack(self.decoder_inputs), tf.float32)
            if not forward_only:
                helper = seq2seq.TrainingHelper(inputs = target_embedded_chars, sequence_length = self.decoder_seq_len, time_major=True)
                # helper = seq2seq.ScheduledEmbeddingTrainingHelper(inputs = dec_inp,
                #                                                   sequence_length = self.decoder_seq_len,
                #                                                  embedding = W_target_emb,
                #                                                  sampling_probability = half,
                #                                                  time_major=True)

            else:
                helper = seq2seq.GreedyEmbeddingHelper(target_embedded_chars, 
                                                       start_tokens=self.decoder_inputs[0],
                                                       end_token=data_utils.EOS_ID)

        
            #, 
                                 #embedding_classes=target_vocab_size, embedding_size=size)
            #DeviceWrapper omitted
            my_decoder = seq2seq.BasicDecoder(cell=MultiRNNCell([DeviceWrapper(ResidualWrapper(LSTMBlockCell(num_units=size)),
                                                                               device='/gpu:%d' % i) for i in range(self.num_gpus) ]),
                                              helper=helper,
                                              initial_state=encoder_final_state)
        
            decoder_outputs, decoder_state = seq2seq.dynamic_decode(my_decoder, output_time_major=False, parallel_iterations=32,
                       swap_memory = True)

            #decoder = seq2seq.AttentionDecoder(cell=cell, helper=helper,attention=tf.contrib,initial_state = encoder_final_state)
            #decoder_outputs, decoder_state = dynamic_decode(decoder, output_time_major=True, parallel_iterations=32,
            #               swap_memory = True)

    with tf.name_scope("loss"):
        #import ipdb
        #ipdb.set_trace()
        self.logits = tf.reshape(decoder_outputs.rnn_output, [self.batch_size, self.decoder_max_size, -1], name="logits_")
        self.sample_ids = decoder_outputs.sample_id
        self.loss = loss.sequence_loss(self.logits, tf.transpose(tf.stack(targets), [1,0], name="targets_"),
                                                     tf.transpose(tf.stack(self.target_weights), [1,0], name="weights_"),
                                                     softmax_loss_function = softmax_loss_function)
`
